source/expectations/data_values/ut_data_value_anydata.tpb

```diff
@@ -16,30 +16,65 @@ create or replace type body ut_data_value_anydata as
   limitations under the License.
   */
 
-  constructor function ut_data_value_anydata(self in out nocopy ut_data_value_anydata) return self as result is
-  begin
-    self.self_type  := $$plsql_unit;
-    self.data_type  := 'undefined';
-    return;
-  end;
-
   overriding member function is_null return boolean is
   begin
     return true;
   end;
 
   overriding member function to_string return varchar2 is
+    l_result varchar2(32767);
+    l_clob   clob;
   begin
-    return ut_utils.to_string( to_char(null) );
+    if self.is_null() then
+      l_result := ut_utils.to_string( to_char(null) );
+    else
+      ut_expectation_processor.set_xml_nls_params();
+      select xmlserialize(content xmltype(self.data_value) indent) into l_clob from dual;
+      l_result := ut_utils.to_string( l_clob, null );
+      ut_expectation_processor.reset_nls_params();
+    end if;
+    return self.format_multi_line( l_result );
+  end;
+
+  overriding member function compare_implementation(a_other ut_data_value) return integer is
+    l_self_data  xmltype;
+    l_other_data xmltype;
+    l_other  ut_data_value_anydata;
+    l_result integer;
+    procedure exclude_xpaths(a_xml in out nocopy xmltype, a_xpath varchar2) is
+    begin
+      select case when a_xpath is not null then deletexml( a_xml, a_xpath ) else a_xml end into a_xml from dual;
+    end;
+  begin
+    if a_other is of (ut_data_value_anydata) then
+      l_other := treat(a_other as ut_data_value_anydata);
+      --needed for 11g xe as it fails on constructing XMLTYPE from null ANYDATA
+      if not self.is_null() and not l_other.is_null() then
+        ut_expectation_processor.set_xml_nls_params();
+        l_self_data := xmltype.createxml(self.data_value);
+        l_other_data := xmltype.createxml(l_other.data_value);
+        exclude_xpaths(l_self_data, coalesce(self.exclude_xpath, l_other.exclude_xpath));
+        exclude_xpaths(l_other_data, coalesce(self.exclude_xpath, l_other.exclude_xpath));
+        ut_expectation_processor.reset_nls_params();
+        if l_self_data is not null and l_other_data is not null then
+          l_result := dbms_lob.compare( l_self_data.getclobval(), l_other_data.getclobval() );
+        end if;
+      end if;
+    else
+      raise value_error;
+    end if;
+    return l_result;
   end;
 
-  overriding member function compare_implementation( a_other ut_data_value ) return integer is
+  final member procedure init(self in out nocopy ut_data_value_anydata, a_value anydata, a_self_type varchar2) is
```

Comment:
How can they introduce new implementations? Open-close principle. I believe it's not final